### ISSUE TITLE
Update: Throughout - Format Function Arguments

### ISSUE THREAD

**Issue by wesbland** [mpi-forum/mpi-issues#249]
## Problem

Do we use MPI_FUNC(all args) in text, or just MPI_FUNC? Section 3.8 uses MPI_IMPROBE(source, tag, ...)

### Suggested Fix

Need to discuss with chapter authors

**Comment by wesbland** [comment]
Need discussion about solution, then chapter committees will need implement.

* Editor - 3
* Chapter committee - 4
* Errata - 0
* Full Proposal - 0

**Comment by bosilca** [comment]
I think having named arguments in the text for some functions make perfect sense, especially when we use the argument name to explain the behavior of the function. Specifically in the pt2pt chapter we explicitly named the arguments for iprobe and improbe, and this is helpful to understand the relationship between the matching requirements behind this call and the returned value of the flag.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
The arguments are named in the LIS binding. If you need a variable name then use a code example. It is valid to use \mpiarg{thing} in the text relying only on the function binding, IMHO.

**Comment by bosilca** [comment]
Sure, but that prototype can be on another page. Having the arguments in the text make the paragraph self-consistent and self-explanatory.

**Comment by martinruefenacht** [comment]
I agree with both, i.e. don't care which one.

I have gone through pt2pt & topol and found the different "styles":
- MPI_FUNC
- MPI_FUNC()
- MPI_FUNC(all, parameter, names), ironically humans at work, names are not correct sometimes 
- MPI_FUNC(..., some, parameters), other variations would be (name, ...) or (..., name, ...)

These were used with inline examples or generally examples:
- MPI_FUNC(C code, and, arguments, mixed), "MPI_WAIT(&array_of_requests[i], status)"
- MPI_FUNC MPI_ARG(C code), "\mpifunc{MPI\_START} \mpiarg{(\&array\_of\_requests[i])}"
- MPI_FUNC(names, constants), "MPI\_DIMS\_CREATE(6, 2, dims)"

1. we need to determine which styles are allowed
2. write tests
    * find style violations
    * find naming violations, function & parameter verification
3. solve the problems present
     * Do all this by hand until it is correct and passes all tests, which is probably not horrific probably easier than pythonization (which would still requiring editing everywhere)
         * most references are MPI_FUNC, so if that is the allowed style which is likely then no-op
    * Pythonize lookup, auto-generate inline accepted styling which makes it a global choice
        * several accessors required since we want to have "the function", "the function with these arguments", "the function with this example code/constant"
        * problems remain with lookup need to be correct reference, at least early failure
            * I am looking into pretty "DSL" with ENUMs instead of string references allowing nicer access, this is also true for in mpi-binding code

**Comment by wgropp** [comment]
I prefer the short form (MPI_FUNC) and to ensure that references to parameter names match those in the "funcdef" definition. I find the other formats distracting and frankly ugly.

**Comment by RolfRabenseifner** [comment]
For me, it looks not so complicated:

We have an existing and voted MPI-3.1 and MPI-4.0 text.
We have some latex macros to make this text better and faster readable.
And the use of these macros is sometimes not done as expected by the writer of the macro.

This issue should be only about formatting because any real text change would be outside of the scope of MPI-4.0 RC, editor pass and very late chapter committee changes after the complete 4.0 RC review phase. 

We have mainly the following usecases of \mpifunc:
- \mpifunc{MPI\_XXXX}
- \mpifunc{MPI\_XXXX(arg1, arg2, ...)}

We can change this by a general editing process into
- \mpifunc{MPI\_XXXX}
- \mpifunc{MPI\_XXXX}\mpiarg{(arg1, arg2, ...)}

There are some advantages regarding the indexing, but I first want to learn whether there would be a formatting drawback?
- \mpifunc{MPI\_XXXX(arg1, arg2, ...)} translates (with ignoring the indexing) into
  \protect\gb\protect\textsf{MPI\_XXXX(arg1, arg2, ...)}
- \mpifunc{MPI\_XXXX}\mpiarg{(arg1, arg2, ...)} translates into (if I understand latex correctly)
  \protect\gb\protect\textsf{MPI\_XXXX}\protect\gb\textsf{(arg1, arg2, ...)}

Question to Bill and all latex specialists: Does this make any difference in the pdf file, e.g., regarding linebreaks?

If there is no difference,m then lets do, but with a clear diff of the pdf files before and after, because there should not be any diff.

**Comment by wgropp** [comment]
There’s only a small difference (the extra \gb before the open paren in the second form makes a linebreak more likely at that position, but only if the line is hard to format without large spaces). However, \mpiarg should be used only with a single argument - its not intended for a parameter list. Better would be to define \mpifuncparms with

\newcommand{\mpifuncparms}[1]{\IImpifuncfmt{#1}}

Bill

William Gropp
Director, NCSA
Thomas M. Siebel Chair in Computer Science
University of Illinois Urbana-Champaign
IEEE-CS President-Elect

**Comment by martinruefenacht** [comment]
Since nobody else has chimed in I will start implementing my understanding of the above discussions. Summary below:

- \mpifunc{MPI_XXX} preferred, unless parameters required by context
- \mpifunc{MPI_XXX()}, **not allowed**
- \mpifunc{MPI_XXX}\mpifuncparams{(..., parameter, parameter, ...)}, minimal required set for context, ellipsis to be used for front missing or back missing parameter names which are not required, or middle?
- \mpifunc{MPI_XXX}\mpifuncparams{(all, parameter, names)}, **not allowed**

This will necessitate some changes, but are purely visual not removing contextual information which has been approved and reviewed.

I will use your suggested \mpifuncparams macro above. I tested the binding chapter and it seems that the function references look identical to MPI 3.1.

The only ambiguous aspect is how the choice for the ellipses should be done. Depending on which parameters names happen to be used we could end up with something like (..., name1, ..., name2, ...), which seems needlessly complicated. Perhaps just scrap the idea of ellipses and require the list of relevant parameter names?

 @wgropp Do you think these changes are sane? What is your opinion on the ellipses?

I will write a set of tests which will look for all cases and throw a fit if someone tries to violate the style for function referencing (regression tests).

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Note of caution about rule 4: some MPI procedures have very few parameters; the minimal set of arguments permitted in rule 3 may supply values for all parameters, which is disallowed by rule 4.

Note of unsolicited opinion: there is *always* a way to re-word the sentence to avoid the contextual 'need' for a (partial) parameter/argument list. That may, or may not, involve individually referencing the argument/parameter names instead. For example, see Issue https://github.com/mpi-forum/mpi-issues/issues/190 and PR https://github.com/mpi-forum/mpi-standard/pull/225. The pattern given in Rule 3 should be "not allowed" as well.

(Here I'm using the convention that procedures define parameters, whereas programs pass arguments to supply values for those parameters. This convention suggests that the `mpiarg` macro should actually be called `mpiparam`, but I digress.)

**Comment by martinruefenacht** [comment]
I modified rule 4 to include context sensitivity. Yes, if all parameters are discussed then all parameters should be presented.

I agree about rule 3, rewording would be possible, but given Rolf's objection to late changes of this magnitude I am happy to kick it down the road. I just want to get this last issue fixed within reason before next week's deadline.

I'll let you spar with global substitutions with respect to `mpiparam`.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Personally, I am currently of the opinion that this should be postponed to MPI-4.1 where we can take the necessary time and votes to re-word all these instances and avoid the argument/parameter list problem entirely. At that point, someone might volunteer to write a (simpler) rule that disallows the presence of an argument/parameter list in the text.

**Comment by martinruefenacht** [comment]
> Personally, I am currently of the opinion that this should be postponed to MPI-4.1 where we can take the necessary time and votes to re-word all these instances and avoid the argument/parameter list problem entirely. At that point, someone might volunteer to write a (simpler) rule that disallows the presence of an argument/parameter list in the text.

I am fine with that as well. Who has such authority? @wesbland @wgropp

**Comment by wgropp** [comment]
Yes, I would like to defer these changes until MPI 4.1. It is important to carefully consider updates to the macros and their use - otherwise we run into the issue of well-intentioned but inconsistent and incompatible changes to the macros, which cause more problems than they fix. 

My preference is to continue to enhance the macros so that they properly label their arguments. What Martin suggests is inline with this, and I think the same approach applies to the other binding macros (which ages ago were defined without thought to such further processing). This is important for the automatic indexing, which is the only practical way to handle references to MPI items in the text, and also gives us a better method to assemble the summaries of names in the appendix (e.g., we can automate the formation of the list of info keys and values).
